frontend/src/data/parseReportData.ts

```diff
@@ -0,0 +1,192 @@
+export const parseReportData = (lines: any[]): ReportData => {
+  const runInfoData = lines[0] as RunInfo;
+  const implementationEntries = Object.entries(runInfoData.implementations) as [
+    string,
+    ImplementationMetadata,
+  ][];
+
+  implementationEntries.sort(([id1], [id2]) => id1.localeCompare(id2));
+  const caseMap = new Map() as Map<number, Case>;
+  const implementationMap = new Map() as Map<string, ImplementationData>;
+  implementationEntries.forEach(([id, metadata]) =>
+    implementationMap.set(id, {
+      id,
+      metadata,
+      cases: new Map(),
+      erroredCases: 0,
+      skippedTests: 0,
+      unsuccessfulTests: 0,
+      erroredTests: 0,
+    }),
+  );
+
+  let didFailFast = false;
+  for (let i = 1; i < lines.length; i++) {
+    const line = lines[i];
+
+    if (line.case) {
+      caseMap.set(line.seq, line.case as Case);
+    } else if (line.implementation) {
+      const caseData = caseMap.get(line.seq)!;
+      const implementationData = implementationMap.get(line.implementation)!;
+      if (line.caught !== undefined) {
+        const errorMessage = line.context?.message ?? line.context?.stderr;
+        implementationData.erroredCases++;
+        implementationData.erroredTests += caseData.tests.length;
+        implementationData.cases.set(
+          line.seq,
+          new Array(caseData.tests.length).fill({
+            state: "errored",
+            message: errorMessage,
+          }),
+        );
+      } else if (line.skipped) {
+        implementationData.skippedTests += caseData.tests.length;
+        implementationData.cases.set(
+          line.seq,
+          new Array(caseData.tests.length).fill({
+            state: "skipped",
+            message: line.message,
+          }),
+        );
+      } else if (line.implementation) {
+        const caseResults = (line.results as any[]).map<CaseResult>(
+          (res, idx) => {
+            if (res.errored) {
+              const errorMessage = res.context?.message ?? res.context?.stderr;
+              implementationData.erroredTests++;
+              return { state: "errored", message: errorMessage };
+            } else if (res.skipped) {
+              implementationData.skippedTests++;
+              return { state: "skipped", message: res.message };
+            } else {
+              const successful = res.valid === caseData.tests[idx].valid;
```

Comment:
If `successful` resolves to true, then we have two cases...
- _**successfully ran**_ and its value is `true` (which is represented by `check` svg)
- _**successfully ran**_ and its value is `false` (which is represented by `cross` SVG)